pkg/shared/utils/arrays/generic_array_utils.go

```diff
@@ -1,12 +1,86 @@
 package arrays
 
+const NotFound = -1
+
+type PredicateFunc[K any] func(x K) bool
+type ReducerFunc[K any, V any] func(accumulatorValue V, currentValue K) V
+type MapperFunc[K any, V any] func(x K) V
+type ConsumerFunc[K any] func(x K)
+
 // FindFirst Finds the first element into the passed in values where the predicate is true
-// WARNING: passed in values are converted to an []interface{} slice and returned value will be an interface{}
-func FindFirst(predicate func(x interface{}) bool, values ...interface{}) (int, interface{}) {
+func FindFirst[K any](values []K, predicate PredicateFunc[K]) (int, K) {
 	for i, val := range values {
 		if predicate(val) {
 			return i, val
 		}
 	}
-	return -1, nil
+	var res K
+	return NotFound, res
+}
+
+// Filter return a slice containing only the values of `values` matching the given predicate
+func Filter[K any](values []K, predicate PredicateFunc[K]) []K {
+	var res []K
+	for _, val := range values {
+		val := val
+		if predicate(val) {
+			res = append(res, val)
+		}
+	}
+	return res
+}
+
+// Map - creates a new slice where each element is the result of the mapper function to the corresponding element in the input slice
+func Map[K any, V any](values []K, mapper MapperFunc[K, V]) []V {
```

Comment:
Should we add the `comparable` constraint to `K` and/or in the `MapperFunc` type?

Reply:
No. The mapper function receives an object and returns another object. The received object does not need to be comparable.